### PR TITLE
Merge different Resource implementation classes

### PR DESCRIPTION
**Pull request by erikjohnston**
We currently have two `Resource` classes that work in slightly different ways, this is an attempt to de-duplicate some of the code and make things more consistent (ironically by adding more classes).

Concretely, this fixes a bug where `trace_servlet` would trigger a `"Tried to close a non-active scope!"` error each time a request was processed in `DirectServeResource` if opentracing is enabled. This is because the `trace_servlet` wrapper was added *after* the `wrap_async_request_handler`, which meant that logcontexts would be forcibly dropped before the handling in `trace_servlet` happened. This is handled correctly by `JsonResource`, so by refactoring things so that they share the same code we can fix that bug.

### CODE REVIEW

File: synapse/rest/key/v2/remote_key_resource.py [hunk 1]
```diff
@@ -20,17 +20,13 @@
 
 from synapse.api.errors import Codes, SynapseError
 from synapse.crypto.keyring import ServerKeyFetcher
-from synapse.http.server import (
-    DirectServeResource,
-    respond_with_json_bytes,
-    wrap_json_request_handler,
```
Comment (clokep) [inline synapse/rest/key/v2/remote_key_resource.py:26]: `wrap_json_request_handler` seems to only be used in one place now (`AdditionalResource`)... perhaps that can be a `DirectServerJsonResource`?

File: synapse/rest/media/v1/thumbnail_resource.py [hunk 2]
```diff
@@ -47,7 +43,6 @@ def __init__(self, hs, media_repo, media_storage):
         self.server_name = hs.hostname
         self.clock = hs.get_clock()
```
Comment (clokep) [inline synapse/rest/media/v1/thumbnail_resource.py:44]: Is this another instance of `self.clock` that can be removed?

File: synapse/http/server.py [hunk 3]
```diff
@@ -383,41 +388,115 @@ def _get_handler_for_request(self, request):
...
+        """Implements _AsyncResource._send_response
+        """
         # TODO: Only enable CORS for the requests that need it.
         respond_with_json(
             request,
             code,
-            response_json_object,
+            response_object,
             send_cors=True,
-            response_code_message=response_code_message,
```
Comment (clokep) [inline synapse/http/server.py:394]: I looked briefly and I suspect that this is unused everywhere? Might be worth checking as a follow-up.

Comment (erikjohnston) [reply synapse/http/server.py:394]: Oh yes, sorry, it is entirely unused.

File: synapse/http/server.py [hunk 4]
```diff
@@ -254,7 +209,89 @@ def register_paths(self, method, path_patterns, callback):
...
+            if isinstance(raw_callback_return, (defer.Deferred, types.CoroutineType)):
+                callback_return = await raw_callback_return
+            else:
+                callback_return = raw_callback_return
+
+            if callback_return is not None:
+                code, response = callback_return
+                self._send_response(request, code, response)
+        except Exception:
+            f = failure.Failure()
```
Comment (richvdh) [inline synapse/http/server.py:258]: probably needs a comment to stop someone "optimising" this to just use the caught exception

**Comment by erikjohnston** [comment]
I think that's basically what we have today, modulo abusing the `@wrap_*_request_handler` to choose between the HTML vs JSON formatting. AFAICT the only code that `JsonServletResource` and `AsyncResource` would share would be the two line `render(..)` function that delegates to `_async_render` (if that)? The `JsonServletResource` doesn't really want to be looking for `async_render_*` and co (though I suppose there is no harm in it).

I guess ideally all I want is to separate the routing from the rendering, although this PR gets a bit unstuck trying to support HTML vs JSON. One possible way of doing this is to have an `SynapseResource` that does the async stuff as well as `trace_servlet` etc, then that just delegates to a router that is passed in as an arg (i.e. it just calls `await self.router.render(request)`). Then the difference between `JsonResource` and `DirectServeResource` is just what router that is passed into `SynapseResource`.

Now, the problem with that is the error handling, as now `SynapseResource` no longer knows the format involved (unless you tell it, in which case you're effectively back to where we are with this PR, modulo  a bit of shuffling). At which point all it can do is kill the request connection if an exception happens that the router(?) doesn't handle (which feels like a small step backwards from where we are today).

> I think you can also hook whatever it is `trace_servlet` is supposed to be doing at `AsyncResource`? Though I must admit I've failed to grok the bug you're trying to fix here.

TLDR: the current code ends up effectively calling `preserve_fn(trace_servlet(callback))` for `JsonResource` but `trace_servlet(preserve_fn(callback))` for `DirectServeResource`. This breaks as ` trace_servlet` assumes the given function correctly follows log context rules. This stems from the fact that the `@wrap_*_request_handler` are put at different levels in the two different classes

**Comment by richvdh** [comment]
> AFAICT the only code that JsonServletResource and AsyncResource would share would be the two line render(..) function that delegates to _async_render (if that)?

There's a bunch of stuff that `AsyncResource` can perform that will be useful for any subclass:

 * `preserve_fn`
 * `request.processing`
 * `trace_request`
 * call `_async_render`
 * return `NOT_DONE_YET`
 * call one of `_handle_error` or `_handle_result`

I call that at least six lines :)

`AsyncResource._async_render` can be the thing that goes looking for `async_render_*` and co, so could be shortcut by subclasses which know they don't need to delegate to it, BUT if we can do this in a way that fixes https://github.com/matrix-org/synapse/issues/6746, that would be good.

> One possible way of doing this is to have an SynapseResource that does the async stuff as well as trace_servlet etc, then that just delegates to a router that is passed in as an arg (i.e. it just calls await self.router.render(request)

I'm not sure I love this plan.

**Comment by richvdh** [comment]
> I call that at least six lines :)

It's also annoying boilerplatey boilerplate, to get all the exception handling right.

**Comment by erikjohnston** [comment]
I am a bit confused how your classes are different than mine, in terms of avoiding mixing of concerns. Each layer seems to be doing pretty much the same things, just implementing things a bit differently (or having `JsonServletResource` inherit from `AsyncJsonResource`).

I think having the render function call out to a function that subclasses implement to handle the request is nicer than it calling a function to get the callback is a nicer way of doing things. (Though its worth noting that doing it like that means we call `trace_servlet(callback)` each request, rather than wrapping the callbacks on startup).

I don't really understand why we would implement stub functions for `_handle_error` and `_handle_result`, everything has a concrete implementation of those already that produce better results.

> `AsyncResource._async_render` can be the thing that goes looking for `async_render_*` and co, so could be shortcut by subclasses which know they don't need to delegate to it, BUT if we can do this in a way that fixes #6746 (and maybe #6894), that would be good.

I think we can make that work either way TBH.

> > One possible way of doing this is to have an SynapseResource that does the async stuff as well as trace_servlet etc, then that just delegates to a router that is passed in as an arg (i.e. it just calls await self.router.render(request)
> 
> I'm not sure I love this plan.

Fair, it was an attempt to try and separate out the concerns between rendering and routing more explicitly.

**Comment by richvdh** [comment]
> I think having the render function call out to a function that subclasses implement to handle the request is nicer than it calling a function to get the callback is a nicer way of doing things

This sentence doesn't make any sense, which is a shame because on closer inspection I think it's basically the nub of the difference between our approaches.

In my approach, rather than having `AsyncResource` call `_get_handler_for_request`, it would instead look for `_async_render*` methods itself - thus removing the point of `DirectRenderResource`. I think you're saying that you prefer one approach over the other, but I can't really tell which you prefer.

> (Though its worth noting that doing it like that means we call trace_servlet(callback) each request, rather than wrapping the callbacks on startup).

trace_servlet doesn't *need* to be implemented as a wrapper... you could do what it's doing inline instead.

> I don't really understand why we would implement stub functions for _handle_error and _handle_result, everything has a concrete implementation of those already that produce better results.

fair, scratch that part of the plan; keep them as abstract functions.

**Comment by richvdh** [comment]
> Fair, it was an attempt to try and separate out the concerns between rendering and routing more explicitly.

Incidentally, I think this would be better solved by using the resource tree as it is intended (ie, having individual `Resource`s for each REST endpoint, which are registered up front), and thus fixing https://github.com/matrix-org/synapse/issues/5118 into the bargain. This whole business where the rendering layer calls out to a routing layer rather than the other way around feels wrong. I don't suggest we attempt to completely change all that now though.

**Comment by erikjohnston** [comment]
I think it would be helpful if you could answer what you're objections are to the current approach that you're trying to fix. 

I do think that your `AsyncResource` is nicer than the one in the PR, but these seem mainly to be cosmetic changes rather than fixing any mixing of concerns.

**Comment by richvdh** [comment]
Two things, basically:

* your `_AsyncResource` is mixing together (1) the handling of async things, and (2) the handling of servlet path registration. This is what I'm saying about mixing concerns.
* Support for `_async_render_METHOD` methods should be in `_AsyncResource`, largely for symmetry with `Resource` itself; I'm not convinced that looking for the methods during construction and sticking them in a dict is a win.

**Comment by erikjohnston** [comment]
Ok, thanks. I guess I just see it all as routing, but am happy to rejig.

**Comment by erikjohnston** [comment]
The problem with the new approach is that now we need to move the `trace_servlet` and `request.metrics.name` into the sub classes since servlet name depends on the routing. At which point we're back to square one where we're duplicating the code.

I suppose the correct, ish, way of fixing this is to have each servlet be a `Resource`, or something, but that is not a change I'm going to make now.

Thoughts on how to get out of this welcome, though at this point I might just give up on the idea of de-duplicating the code and just fix the specific case of opentracing and move on

**Comment by richvdh** [comment]
> The problem with the new approach is that now we need to move the trace_servlet and request.metrics.name into the sub classes since servlet name depends on the routing.

oh, *balls.*

`request.metrics.name` isn't so bad because you can update it later (provided you get there before the first `await`), but `trace_servlet` is indeed problematic. Is it practical to update the span in some way after it is started (add the servlet name as some sort of metadata instead?)

**Comment by richvdh** [comment]
(otherwise yes, that is a major flaw in my plan, and it's up to you whether you stick with your proposal or patch the current situation up)

**Comment by erikjohnston** [comment]
Aha, actually looks like we *can* change the operation name, which isn't possible using the standard opentracing APIs but is possible in the jaeger specific APIs. In which case we can actually just set the operation name based on the request name at the end of the rendering, rather than duplicating that logic.

**Comment by richvdh** [comment]
>  In which case we can actually just set the operation name based on the request name at the end of the rendering

to confirm: this won't confuse nested spans that have come and gone by the time we finish rendering?

**Comment by erikjohnston** [comment]
> > In which case we can actually just set the operation name based on the request name at the end of the rendering
> 
> to confirm: this won't confuse nested spans that have come and gone by the time we finish rendering?

I don't believe so, as they all use opaque IDs internally. Will test on jki.re though.

**Comment by erikjohnston** [comment]
Ok, I've rejigged everything as discussed. Sorry that it's all in a single commit

File: synapse/http/server.py [hunk 5]
```diff
@@ -61,100 +62,54 @@
...
+MaybeAwaitable = Union[Awaitable[T], T]
 
-    The handler method must have a signature of "handle_foo(self, request)",
-    where "request" must be a SynapseRequest.
 
-    The handler must return a deferred or a coroutine. If the deferred succeeds
-    we assume that a response has been sent. If the deferred fails with a SynapseError we use
-    it to send a JSON response with the appropriate HTTP reponse code. If the
-    deferred fails with any other type of error we send a 500 reponse.
+def return_json_error(f: failure.Failure, request: Request) -> None:
```
Comment (richvdh) [inline synapse/http/server.py:75]: still think this needs to be a `SynapseRequest`

File: synapse/http/server.py [hunk 6]
```diff
@@ -79,82 +80,54 @@ def wrap_json_request_handler(h):
...
+            "Failed handle request via %r: %r",
+            request.request_metrics.name,
+            request,
+            exc_info=(f.type, f.value, f.getTracebackObject()),
+        )
 
-    The handler method must have a signature of "handle_foo(self, request)",
-    where "request" must be a SynapseRequest.
-    """
+    if request.startedWriting:
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/http/server.py:95]: still needs a comment

File: synapse/http/server.py [hunk 7]
```diff
@@ -79,82 +80,54 @@ def wrap_json_request_handler(h):
...
 
-    async def wrapped_request_handler(self, request):
-        try:
-            await h(self, request)
-        except Exception:
-            f = failure.Failure()
-            return_html_error(f, request, HTML_ERROR_TEMPLATE)
 
-    return wrap_async_request_handler(wrapped_request_handler)
+TV = TypeVar("TV")
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/http/server.py:112]: still not used

File: synapse/http/server.py [hunk 8]
```diff
@@ -254,7 +209,101 @@ def register_paths(self, method, path_patterns, callback):
         pass
 
 
-class JsonResource(HttpServer, resource.Resource):
+class _AsyncResource(resource.Resource, metaclass=abc.ABCMeta):
+    """Base class for resources that have async handlers.
+
```
Comment (richvdh) [inline synapse/http/server.py:214]: a word or two about how it works and what the subclasses are meant to override would be helpful

File: synapse/http/server.py [hunk 9]
```diff
@@ -254,7 +209,101 @@ def register_paths(self, method, path_patterns, callback):
...
+    def render(self, request):
+        """ This gets called by twisted every time someone sends us a request.
+        """
+        defer.ensureDeferred(self._async_render_wrapper(request))
+        return NOT_DONE_YET
+
+    @wrap_async_request_handler
+    async def _async_render_wrapper(self, request):
+        """This is a wrapper that delegates to `_async_render`,
```
Comment (richvdh) [inline synapse/http/server.py:233]: probably needs more words? or at least the dangling comma fixing

File: synapse/logging/opentracing.py [hunk 10]
```diff
@@ -793,48 +793,41 @@ def _tag_args_inner(*args, **kwargs):
     return _tag_args_inner
 
 
-def trace_servlet(servlet_name, extract_context=False):
-    """Decorator which traces a serlet. It starts a span with some servlet specific
-    tags such as the servlet_name and request information
+@contextlib.contextmanager
+def trace_servlet(request: "SynapseRequest", extract_context: bool = False):
+    """Context manager which traces a servlet. It starts a span with some
+    servlet specific tags such as the servlet_name and request information.
```
Comment (richvdh) [inline synapse/logging/opentracing.py:799]: I don't think this is accurate. Particularly, it does nothing with the servlet_name. Generally I don't think it traces a servlet (and could do with a rename): perhaps "traces a request"?

Comment (richvdh) [reply synapse/logging/opentracing.py:799]: also: aiui a "context manager" is a thing you can call `with` on... so this is a method that *returns* a context manager, rather than being one itself.

File: synapse/logging/opentracing.py [hunk 11]
```diff
@@ -793,48 +793,41 @@ def _tag_args_inner(*args, **kwargs):
...
-                scope = start_active_span_from_request(
-                    request, servlet_name, tags=request_tags
-                )
-            else:
-                scope = start_active_span(servlet_name, tags=request_tags)
-
-            with scope:
-                result = func(request, *args, **kwargs)
+    if opentracing is None:
+        yield
```
Comment (richvdh) [inline synapse/logging/opentracing.py:808]: kinda wondering about the overhead here. There's a `_NullContextManager` in `generic_worker.py` which we could pull out and use...

File: synapse/logging/opentracing.py [hunk 12]
```diff
@@ -793,48 +793,41 @@ def _tag_args_inner(*args, **kwargs):
...
+        scope = start_active_span_from_request(request, request_name, tags=request_tags)
+    else:
+        scope = start_active_span(request_name, tags=request_tags)
 
-        return _trace_servlet_inner
+    with scope:
+        try:
+            yield
+        finally:
+            # We set the operation name again in case its changed (which happens
```
Comment (richvdh) [inline synapse/logging/opentracing.py:829]: > again 

do we set it somewhere else?

File: synapse/replication/http/__init__.py [hunk 13]
```diff
@@ -30,7 +30,9 @@
 
 class ReplicationRestResource(JsonResource):
     def __init__(self, hs):
-        JsonResource.__init__(self, hs, canonical_json=False)
+        # We enabling extracting jaeger contexts here as these are internal
```
Comment (richvdh) [inline synapse/replication/http/__init__.py:33]:
```suggestion
        # We enable extracting jaeger contexts here as these are internal
```

File: synapse/http/server.py [hunk 14]
```diff
@@ -300,37 +347,34 @@ def register_paths(
...
         for path_pattern in path_patterns:
             logger.debug("Registering for %s %s", method, path_pattern.pattern)
             self.path_regexs.setdefault(method, []).append(
                 self._PathEntry(path_pattern, callback, servlet_classname)
             )
 
-    def render(self, request):
-        """ This gets called by twisted every time someone sends us a request.
+    def _get_handler_for_request(self, request: SynapseRequest):
+        """Implements _AsyncResource._get_handler_for_request
```
Comment (richvdh) [inline synapse/http/server.py:360]: _AsyncResource._get_handler_for_request no longer exists. Can has proper docstring pls?

File: synapse/http/server.py [hunk 15]
```diff
@@ -300,37 +347,34 @@ def register_paths(
...
 
-    @wrap_json_request_handler
     async def _async_render(self, request):
-        """ This gets called from render() every time someone sends us a request.
-            This checks if anyone has registered a callback for that method and
-            path.
-        """
         callback, servlet_classname, group_dict = self._get_handler_for_request(request)
 
-        # Make sure we have a name for this handler in prometheus.
```
Comment (richvdh) [inline synapse/http/server.py:333]: this was a valid comment I think?

File: synapse/logging/opentracing.py (hunk 12 above)
Comment (erikjohnston) [reply synapse/logging/opentracing.py:829]: Yes, in `start_active_span`

File: synapse/logging/opentracing.py (hunk 11 above)
Comment (erikjohnston) [reply synapse/logging/opentracing.py:808]: A single if statement per request doesn't fill me with terror, if I'm honest.

**Comment by erikjohnston** [comment]
> have you checked that OPTIONS works everywhere we could conceivably have broken it?

Most of the `OPTIONS` are handled by a separate dedicated resource. For the places that handle options via `_async_render_OPTIONS` themselves we have a test for the url preview case and I've manually tested the other two

**Comment by clokep** [comment]
@erikjohnston Looks like I bitrotted this. There are conflicts now! 😢

File: synapse/http/server.py [hunk 16]
```diff
@@ -62,99 +63,41 @@
...
-            return_html_error(f, request, HTML_ERROR_TEMPLATE)
+        logger.error(
+            "Failed handle request via %r: %r",
+            request.request_metrics.name,
+            request,
+            exc_info=(f.type, f.value, f.getTracebackObject()),
+        )
 
-    return wrap_async_request_handler(wrapped_request_handler)
+    if request.startedWriting:
```
Comment (richvdh) [inline synapse/http/server.py:86]: third time of asking: please can I have a comment, sir? why is this condition necessary or helpful?

File: synapse/http/server.py [hunk 17]
```diff
@@ -249,7 +192,110 @@ def register_paths(self, method, path_patterns, callback):
...
+            request.request_metrics.name = self.__class__.__name__
+
+            with trace_servlet(request, self._extract_context):
+                callback_return = await self._async_render(request)
+
+                if callback_return is not None:
+                    code, response = callback_return
+                    self._send_response(request, code, response)
+        except Exception:
+            f = failure.Failure()
```
Comment (richvdh) [inline synapse/http/server.py:232]: please can this have a comment to stop someone "optimising" this to just use the caught exception

File: synapse/logging/opentracing.py (hunk 11 above)
Comment (richvdh) [reply synapse/logging/opentracing.py:809]: it's not the if statement I'm worrying about, but more the infrastructure involved in instantiating a generator and constructing a GeneratorContextManager to wrap it.

I'm inclined to agree it's not worth lying awake at night over though.

File: tests/http/test_additional_resource.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    def __init__(self, config, module_api):
+        pass
+
+    async def handle_request(self, request):
+        respond_with_json(request, 200, {"some_key": "some_value_sync"})
+
+
+class AdditionalResourceTests(HomeserverTestCase):
+    """Very basic tests that `AdditionalResource` works correctly with sync
+    and async handler.s
```
Comment (richvdh) [inline tests/http/test_additional_resource.py:41]:
```suggestion
    and async handlers.
```

File: synapse/http/server.py [hunk 19]
```diff
@@ -338,81 +387,46 @@ async def _async_render(self, request):
...
-        Render the request, using an asynchronous render handler if it exists.
+        # We expect to get bytes for us to write
+        assert isinstance(response_object, bytes)
+        html_bytes = response_object
+
+        request.setResponseCode(200)
+        request.setHeader(b"Content-Type", b"text/html; charset=utf-8")
+        request.setHeader(b"Content-Length", b"%d" % (len(html_bytes),))
+        request.write(html_bytes)
+        finish_request(request)
```
Comment (richvdh) [inline synapse/http/server.py:422]: please can this use `respond_with_html_bytes` ?

(I'm not sure it's actually used? not that that matters.)

File: synapse/http/server.py (hunk 16 above)
Comment (erikjohnston) [reply synapse/http/server.py:86]: Argh, how do I keep missing this. Sorry.

File: synapse/http/server.py (hunk 19 above)
Comment (erikjohnston) [reply synapse/http/server.py:422]: I don't think it is, but I think it *should* be given we seem to copy and paste the code around. I was going to tidy it up but then thought maybe this PR is large enough as it is :)

File: synapse/logging/opentracing.py (hunk 11 above)
Comment (erikjohnston) [reply synapse/logging/opentracing.py:809]: Ah right I see, I forgot that we'd need to generate a context manager object each time

File: synapse/http/server.py (hunk 19 above)
Comment (clokep) [reply synapse/http/server.py]: FTR `respond_with_html_bytes` was actually used in one spot already (in `synapse.rest.client.v1.pusher`). But 👍 on using it here too! The goal was to not have any places where we manually write out HTML.

Comment (richvdh) [reply synapse/http/server.py]: for clarity: I meant to say that I don't think `DirectServeHtmlResource._send_response` is ever (currently) called - not that that makes any difference to that it should use `respond_with_html_bytes`.
